File: go-backend/models/card.go
package models

import (
	"database/sql"
	"log"
	"time"

	"github.com/pgvector/pgvector-go"
)

type Card struct {
	ID         int           `json:"id"`
	CardID     string        `json:"card_id"`
	UserID     int           `json:"user_id"`
	Title      string        `json:"title"`
	Body       string        `json:"body"`
	Link       string        `json:"link"`
	IsDeleted  bool          `json:"is_deleted"`
	CreatedAt  time.Time     `json:"created_at"`
	UpdatedAt  time.Time     `json:"updated_at"`
	ParentID   int           `json:"parent_id"`
	Parent     PartialCard   `json:"parent"`
	Files      []File        `json:"files"`
	Children   []PartialCard `json:"children"`
	References []PartialCard `json:"references"`
	Keywords   []Keyword     `json:"keywords"`
	Tags       []Tag         `json:"tags"`
	Tasks      []Task        `json:"tasks"`
	Embedding  pgvector.Vector
	Entities   []Entity `json:"entities"`
	TagCount   int
	IsPinned   bool `json:"is_pinned"`
}

func ScanCards(rows *sql.Rows) ([]Card, error) {
	var cards []Card

	for rows.Next() {
		var card Card
		if err := rows.Scan(
			&card.ID,
			&card.CardID,
			&card.UserID,
			&card.Title,
			&card.Body,
			&card.Link,
			&card.ParentID,
			&card.CreatedAt,
			&card.UpdatedAt,
			&card.TagCount,
		); err != nil {
			log.Printf(" query full err %v", err)
			return cards, err
		}
		cards = append(cards, card)
	}

	return cards, nil
}

func ScanPartialCards(rows *sql.Rows) ([]PartialCard, error) {
	var cards []PartialCard

	for rows.Next() {
		var card PartialCard
		if err := rows.Scan(
			&card.ID,
			&card.CardID,
			&card.UserID,
			&card.Title,
			&card.ParentID,
			&card.CreatedAt,
			&card.UpdatedAt,
		); err != nil {
			log.Printf("err %v", err)
			return cards, err
		}
		cards = append(cards, card)

	}
	return cards, nil

}

type PartialCard struct {
	ID        int       `json:"id"`
	CardID    string    `json:"card_id"`
	UserID    int       `json:"user_id"`
	Title     string    `json:"title"`
	ParentID  int       `json:"parent_id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	Tags      []Tag     `json:"tags"`
}

func ConvertCardToPartialCard(input Card) PartialCard {
	return PartialCard{
		ID:        input.ID,
		CardID:    input.CardID,
		UserID:    input.UserID,
		Title:     input.Title,
		ParentID:  input.ParentID,
		CreatedAt: input.CreatedAt,
		UpdatedAt: input.UpdatedAt,
		Tags:      input.Tags,
	}
}

type EditCardParams struct {
	CardID string `json:"card_id"`
	Title  string `json:"title"`
	Body   string `json:"body"`
	Link   string `json:"link"`
}

type NextIDParams struct {
	CardType string `json:"card_type"`
}

type NextIDResponse struct {
	Error   bool   `json:"error"`
	Message string `json:"message"`
	NextID  string `json:"new_id"`
}

type CardChunk struct {
	ID               int       `json:"id"`
	CardID           string    `json:"card_id"`
	UserID           int       `json:"user_id"`
	Title            string    `json:"title"`
	Chunk            string    `json:"body"`
	CreatedAt        time.Time `json:"created_at"`
	UpdatedAt        time.Time `json:"updated_at"`
	ParentID         int       `json:"parent_id"`
	Ranking          float64   `json:"ranking"`
	SharedEntities   int       `json:"shared_entities"`
	EntitySimilarity float64   `json:"entity_similarity"`
	CombinedScore    float64   `json:"combined_score"`
}

func ScanCardChunks(rows *sql.Rows) ([]CardChunk, error) {
	var cards []CardChunk

	for rows.Next() {
		var card CardChunk
		if err := rows.Scan(
			&card.ID,
			&card.CardID,
			&card.UserID,
			&card.Title,
			&card.Chunk,
			&card.CreatedAt,
			&card.UpdatedAt,
			&card.ParentID,
			&card.Ranking,
			&card.SharedEntities,
			&card.EntitySimilarity,
			&card.CombinedScore,
		); err != nil {
			return cards, err
		}
		cards = append(cards, card)
	}
	return cards, nil
}

func ConvertCardToChunk(input Card) CardChunk {
	return CardChunk{
		ID:        input.ID,
		CardID:    input.CardID,
		UserID:    input.UserID,
		Title:     input.Title,
		Chunk:     input.Body,
		ParentID:  input.ParentID,
		CreatedAt: input.CreatedAt,
		UpdatedAt: input.UpdatedAt,
	}
}

File: zettelkasten-front/src/models/Card.ts
import { File } from "./File";
import { Tag } from "./Tags";
import { Task } from "./Task";

export interface PartialCard {
  id: number;
  card_id: string;
  user_id: number;
  title: string;
  parent_id: number;
  created_at: Date;
  updated_at: Date;
  tags: Tag[];
}

export interface Entity {
  id: number;
  user_id: number;
  name: string;
  description: string;
  type: string;
  created_at: Date;
  updated_at: Date;
  card_count: number;
  card_pk: number | null;
  card?: PartialCard;
}

export interface Card {
  id: number;
  card_id: string;
  user_id: number;
  title: string;
  body: string;
  link: string;
  is_deleted: boolean;
  created_at: Date;
  updated_at: Date;
  parent_id: number;
  parent: PartialCard;
  files: File[];
  children: PartialCard[];
  references: PartialCard[];
  tags: Tag[];
  tasks: Task[];
  entities: Entity[];
  is_pinned?: boolean; // Whether the current user has pinned this card
}

export interface CardChunk {
  id: number;
  card_id: string;
  user_id: number;
  title: string;
  body: string;
  parent_id: number;
  created_at: Date;
  updated_at: Date;
  tags: Tag[];
  ranking: number;
  combined_score: number;
  shared_entities: number;
  entity_similarity: number;
}

export interface SearchResult {
  id: string;
  type: string;
  title: string;
  preview: string;
  score: number;
  created_at: Date;
  updated_at: Date;
  metadata: {
    parent_id?: number;
    shared_entities?: number;
    entity_similarity?: number;
    semantic_ranking?: number;
    [key: string]: any;
  };
}

export const defaultPartialCard: PartialCard = {
  id: -1,
  card_id: "",
  user_id: -1,
  title: "",
  parent_id: -1,
  created_at: new Date(0),
  updated_at: new Date(0),
  tags: [],
};

export const defaultCard: Card = {
  id: -1,
  card_id: "",
  user_id: -1,
  title: "",
  body: "",
  link: "",
  is_deleted: false,
  created_at: new Date(0),
  updated_at: new Date(0),
  parent_id: -1,
  parent: defaultPartialCard,
  files: [],
  children: [],
  references: [],
  tags: [],
  tasks: [],
  entities: [],
  is_pinned: false,
};

export interface NextIdResponse {
  error: boolean;
  message: string;
  new_id: string;  // Matches the actual backend response
}

export interface FlashcardRecordNextParams {
  card_pk: number;
  rating: number;
}
enum Rating {
  Again = 0,
  Hard = 1,
  Good = 2,
  Easy = 3
}

export function getRatingValue(rating: string): number {
  switch (rating.toLowerCase()) {
    case 'again':
      return Rating.Again;
    case 'hard':
      return Rating.Hard;
    case 'good':
      return Rating.Good;
    case 'easy':
      return Rating.Easy;
    default:
      throw new Error(`Invalid rating value: ${rating}`);
  }
}

export interface CardTemplate {
  id: number;
  user_id: number;
  title: string;
  body: string;
  created_at: Date;
  updated_at: Date;
}

export const defaultCardTemplate: CardTemplate = {
  id: -1,
  user_id: -1,
  title: "",
  body: "",
  created_at: new Date(0),
  updated_at: new Date(0),
};
